File: native/ios/Sources/NativeAPI.swift
import Foundation

final class RejectRedirects: NSObject, URLSessionTaskDelegate {
    func urlSession(_ session: URLSession, task: URLSessionTask, willPerformHTTPRedirection response: HTTPURLResponse,
                    newRequest request: URLRequest, completionHandler: @escaping (URLRequest?) -> Void) { completionHandler(nil) }
}

enum NativeHTTPMethod: String {
    case get = "GET", post = "POST", put = "PUT", delete = "DELETE"
}

// Fixed backend wire outcomes; arbitrary server text is never retained in failures.
enum NativeFailureOutcome {
    case erasureInProgress, erasureRequestChanged, erasureRetryRequired, exportIncomplete

    static func decode(status: Int, data: Data) -> NativeFailureOutcome? {
        guard hasUnambiguousJSONStructure(data) else { return nil }
        guard let value = (try? JSONSerialization.jsonObject(with: data)) as? [String: Any] else { return nil }
        let state = value["status"] as? String
        let code = value["code"] as? String
        if status == 409 && value["code"] == nil {
            if state == "IN_PROGRESS" { return .erasureInProgress }
            if state == "REQUEST_CHANGED" { return .erasureRequestChanged }
        }
        if status == 503 && state == "RETRY_REQUIRED" && code == "ERASURE_INCOMPLETE" { return .erasureRetryRequired }
        if status == 503 && value["status"] == nil && code == "DATA_EXPORT_INCOMPLETE" { return .exportIncomplete }
        return nil
    }

    // Foundation supplies value/escape validation. Reject syntax it can normalize
    // away before decoding: duplicate top-level keys and trailing commas.
    private static func hasUnambiguousJSONStructure(_ data: Data) -> Bool {
        guard String(data: data, encoding: .utf8) != nil else { return false }
        let bytes = Array(data)
        var depth = 0
        var previous: UInt8?
        var stringStart: Int?
        var lastString: Range<Int>?
        var escaped = false
        var names = Set<String>()
        for (index, byte) in bytes.enumerated() {
            if let start = stringStart {
                if byte < 0x20 { return false }
                if escaped { escaped = false }
                else if byte == 0x5c { escaped = true }
                else if byte == 0x22 {
                    stringStart = nil
                    lastString = start..<(index + 1)
                    previous = byte
                }
                continue
            }
            if byte == 0x20 || byte == 0x09 || byte == 0x0a || byte == 0x0d { continue }
            switch byte {
            case 0x22: stringStart = index
            case 0x7b, 0x5b: depth += 1
            case 0x7d, 0x5d:
                if previous == 0x2c { return false }
                depth -= 1
            case 0x3a where depth == 1:
                guard previous == 0x22, let range = lastString,
                      let name = (try? JSONSerialization.jsonObject(
                        with: Data(bytes[range]), options: [.fragmentsAllowed])) as? String,
                      names.insert(name).inserted else { return false }
            default: break
            }
            previous = byte
        }
        return stringStart == nil && depth == 0
    }
}

struct NativeResponse {
    let body: Data
    let status: Int
    let exportIntegrity: String?
}

final class NativeAPI {
    private let config: MobileConfiguration
    private let session: URLSession
    init(config: MobileConfiguration, protocolClasses: [AnyClass]? = nil) {
        self.config = config
        let settings = URLSessionConfiguration.ephemeral
        if let protocolClasses { settings.protocolClasses = protocolClasses }
        settings.urlCache = nil
        settings.httpCookieStorage = nil
        settings.httpShouldSetCookies = false
        settings.requestCachePolicy = .reloadIgnoringLocalCacheData
        settings.timeoutIntervalForRequest = config.requestTimeoutSeconds
        settings.timeoutIntervalForResource = config.requestTimeoutSeconds
        session = URLSession(configuration: settings, delegate: RejectRedirects(), delegateQueue: nil)
    }
    deinit { session.invalidateAndCancel() }

    func request(access: SessionAccess, service: String, path: String, query: [String: String] = [:], body: [String: Any]? = nil, method: NativeHTTPMethod? = nil) async throws -> Data {
        let response = try await requestResponse(access: access, service: service, path: path, query: query, body: body, method: method)
        return response.body
    }

    func requestResponse(access: SessionAccess, service: String, path: String, query: [String: String] = [:], body: [String: Any]? = nil, method: NativeHTTPMethod? = nil) async throws -> NativeResponse {
        let selectedMethod = method ?? (body == nil ? .get : .post)
        switch selectedMethod {
        case .get: guard body == nil else { throw MobileFailure.configuration }
        case .post, .put: guard body != nil else { throw MobileFailure.configuration }
        case .delete: break
        }
        guard access.identity.expiresAt > Date(), let base = config.services[service],
              var url = URLComponents(url: base, resolvingAgainstBaseURL: false),
              path.range(of: "^/[A-Za-z0-9/_.-]+$", options: .regularExpression) != nil,
              !path.contains("//"), !path.split(separator: "/").contains("..") else { throw MobileFailure.forbidden }
        url.path = url.path.trimmingCharacters(in: CharacterSet(charactersIn: "/")) + path
        if !url.path.hasPrefix("/") { url.path = "/" + url.path }
        url.queryItems = query.isEmpty ? nil : query.map { URLQueryItem(name: $0.key, value: $0.value) }
        guard let endpoint = url.url else { throw MobileFailure.configuration }
        var request = URLRequest(url: endpoint)
        request.httpMethod = selectedMethod.rawValue
        if let body {
            request.httpBody = try JSONSerialization.data(withJSONObject: body)
            request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        }
        request.setValue("Bearer " + access.token, forHTTPHeaderField: "Authorization")
        request.setValue(config.residency, forHTTPHeaderField: "x-user-region")
        request.setValue("application/json", forHTTPHeaderField: "Accept")
        request.setValue("no-store", forHTTPHeaderField: "Cache-Control")
        let (bytes, response) = try await session.bytes(for: request)
        defer { bytes.task.cancel() }
        guard let response = response as? HTTPURLResponse else { throw HTTPFailure(status: nil) }
        let successful = (200..<300).contains(response.statusCode)
        var data = Data()
        do {
            guard response.expectedContentLength <= config.maxResponseBytes else { throw MobileFailure.response }
            for try await byte in bytes {
                try Task.checkCancellation()
                guard data.count < config.maxResponseBytes else { throw MobileFailure.response }
                data.append(byte)
            }
        } catch {
            try Task.checkCancellation()
            if !successful { throw HTTPFailure(status: response.statusCode) }
            throw error
        }
        try Task.checkCancellation()
        if !successful {
            throw HTTPFailure(status: response.statusCode,
                              outcome: NativeFailureOutcome.decode(status: response.statusCode, data: data))
        }
        return NativeResponse(body: data, status: response.statusCode,
                              exportIntegrity: response.value(forHTTPHeaderField: "X-Export-Integrity"))
    }
}

File: native/android/app/src/main/java/com/mediconnect/nativeapp/NativeApi.kt
package com.mediconnect.nativeapp

import kotlinx.coroutines.suspendCancellableCoroutine
import okhttp3.Call
import okhttp3.Callback
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.Response
import org.json.JSONObject
import com.google.gson.Strictness
import com.google.gson.stream.JsonReader
import com.google.gson.stream.JsonToken
import java.io.IOException
import java.io.StringReader
import java.nio.ByteBuffer
import java.nio.charset.CodingErrorAction
import java.nio.charset.StandardCharsets
import java.time.Instant
import kotlin.coroutines.resume
import kotlin.coroutines.resumeWithException

import okhttp3.MediaType.Companion.toMediaType
import okhttp3.RequestBody.Companion.toRequestBody

enum class NativeHttpMethod { GET, POST, PUT, DELETE }

// Fixed backend wire outcomes; arbitrary server text is never retained in failures.
enum class NativeFailureOutcome {
    ERASURE_IN_PROGRESS, ERASURE_REQUEST_CHANGED, ERASURE_RETRY_REQUIRED, EXPORT_INCOMPLETE;

    companion object {
        fun decode(status: Int, body: String): NativeFailureOutcome? = try {
            // Use the same strict reader on the JVM and Android. JSONObject accepts
            // non-JSON syntax and silently ignores trailing input on some runtimes.
            val value = mutableMapOf<String, String?>()
            JsonReader(StringReader(body)).use { reader ->
                reader.strictness = Strictness.STRICT
                reader.beginObject()
                while (reader.hasNext()) {
                    val name = reader.nextName()
                    require(!value.containsKey(name))
                    value[name] = if (reader.peek() == JsonToken.STRING) reader.nextString()
                    else { reader.skipValue(); null }
                }
                reader.endObject()
                require(reader.peek() == JsonToken.END_DOCUMENT)
            }
            val state = value["status"]
            val code = value["code"]
            when {
                status == 409 && !value.containsKey("code") && state == "IN_PROGRESS" -> ERASURE_IN_PROGRESS
                status == 409 && !value.containsKey("code") && state == "REQUEST_CHANGED" -> ERASURE_REQUEST_CHANGED
                status == 503 && state == "RETRY_REQUIRED" && code == "ERASURE_INCOMPLETE" -> ERASURE_RETRY_REQUIRED
                status == 503 && !value.containsKey("status") && code == "DATA_EXPORT_INCOMPLETE" -> EXPORT_INCOMPLETE
                else -> null
            }
        } catch (_: Exception) { null }
    }
}

class NativeResponse(body: ByteArray, val status: Int, val exportIntegrity: String?) {
    private val retainedBytes = body.copyOf()
    val body: String get() = retainedBytes.toString(StandardCharsets.UTF_8)
    val bytes: ByteArray get() = retainedBytes.copyOf()
}

class NativeApi(private val config: MobileConfiguration, private val sessions: SessionProvider, private val client: OkHttpClient) {
    suspend fun request(identity: Identity, service: String, path: String, query: Map<String, String> = emptyMap(), body: JSONObject? = null, method: NativeHttpMethod? = null): String =
        requestResponse(identity, service, path, query, body, method).body

    suspend fun requestResponse(identity: Identity, service: String, path: String, query: Map<String, String> = emptyMap(), body: JSONObject? = null, method: NativeHttpMethod? = null): NativeResponse {
        val selectedMethod = method ?: if (body == null) NativeHttpMethod.GET else NativeHttpMethod.POST
        require(when (selectedMethod) {
            NativeHttpMethod.GET -> body == null
            NativeHttpMethod.POST, NativeHttpMethod.PUT -> body != null
            NativeHttpMethod.DELETE -> true
        })
        val session = sessions.fetch()
        require(session.identity.subject == identity.subject && session.identity.role == identity.role && session.identity.expiresAt > Instant.now())
        require(path.matches(Regex("/[A-Za-z0-9/_.-]+")) && !path.contains("//") && !path.split('/').contains(".."))
        val base = config.services[service] ?: throw ApiFailure()
        val url = base.newBuilder().encodedPath(base.encodedPath.trimEnd('/') + path)
        query.forEach { (key, value) -> url.addQueryParameter(key, value) }
        val builder = Request.Builder().url(url.build())
            .header("Authorization", "Bearer ${session.token}").header("x-user-region", config.residency)
            .header("Accept", "application/json").header("Cache-Control", "no-store")
        builder.method(selectedMethod.name, body?.toString()?.toRequestBody("application/json".toMediaType()))
        return execute(client.newCall(builder.build()))
    }

    private suspend fun execute(call: Call): NativeResponse = suspendCancellableCoroutine { continuation ->
        continuation.invokeOnCancellation { call.cancel() }
        call.enqueue(object : Callback {
            override fun onFailure(call: Call, e: IOException) {
                if (continuation.isActive) continuation.resumeWithException(ApiFailure())
            }
            override fun onResponse(call: Call, response: Response) {
                response.use {
                    try {
                        val body = it.body
                        if (body.contentLength() > config.maxResponseBytes) throw ApiFailure()
                        val source = body.source()
                        source.request(config.maxResponseBytes + 1)
                        if (source.buffer.size > config.maxResponseBytes) throw ApiFailure()
                        val bytes = source.readByteArray()
                        val text = if (it.isSuccessful) bytes.toString(StandardCharsets.UTF_8) else StandardCharsets.UTF_8.newDecoder()
                            .onMalformedInput(CodingErrorAction.REPORT)
                            .onUnmappableCharacter(CodingErrorAction.REPORT)
                            .decode(ByteBuffer.wrap(bytes)).toString()
                        if (!it.isSuccessful) throw ApiFailure(it.code, NativeFailureOutcome.decode(it.code, text))
                        val integrity = it.headers.values("X-Export-Integrity").singleOrNull()
                        if (continuation.isActive) continuation.resume(NativeResponse(bytes, it.code, integrity))
                    } catch (failure: Exception) {
                        if (continuation.isActive) continuation.resumeWithException(
                            if (failure is ApiFailure) ApiFailure(it.code, failure.outcome) else ApiFailure(it.code))
                    }
                }
            }
        })
    }
}
